Lib/tkinter/__init__.py

```diff
@@ -763,15 +763,31 @@ def after_cancel(self, id):
 
         Identifier returned by after or after_idle must be
         given as first parameter."""
+        if id is None:
+            return
         try:
-            data = self.tk.call('after', 'info', id)
-            # In Tk 8.3, splitlist returns: (script, type)
-            # In Tk 8.4, splitlist may return (script, type) or (script,)
+            data = self.after_info(id)
```

Comment:
This change requires that tkinter after_info return a reference to the tcl wrapper, (which I think is wrong), so that the wrapper can be deleted (which it should be).  I think after_cancel should continue using the tk.call here.  The only change here might deleting the 8.3 note, which I believe is obsolete, but this is up to Serhiy.

Reply:
It looks to me, that `after info $id` returned a list of 2 elements not in all versions. We can now ignore old unsupported versions, but we should check sources of supported versions. We shouldn't believe the documentation, it contains errors.

Reply:
I've looked at the 8.3 Tcl code (from 2001) and the current (8.6) code.  I'm not sure if the comment about 8.3 and 8.4 was accurate.  My thought is that somehow, `id` was `None` on a call, thus causing `info` to return the list of event_handlers instead of a tuple, as the previous code was expecting.

Here is the 8.3 Tcl code:
```
	case AFTER_INFO: {
	    Tcl_Obj *resultListPtr;

	    if (objc == 2) {
		for (afterPtr = assocPtr->firstAfterPtr; afterPtr != NULL;
		     afterPtr = afterPtr->nextPtr) {
		    if (assocPtr->interp == interp) {
			sprintf(buf, "after#%d", afterPtr->id);
			Tcl_AppendElement(interp, buf);
		    }
		}
		return TCL_OK;
	    }
	    if (objc != 3) {
		Tcl_WrongNumArgs(interp, 2, objv, "?id?");
		return TCL_ERROR;
	    }
	    afterPtr = GetAfterEvent(assocPtr, objv[2]);
	    if (afterPtr == NULL) {
		Tcl_AppendResult(interp, "event \"", Tcl_GetString(objv[2]),
			"\" doesn't exist", (char *) NULL);
		return TCL_ERROR;
	    }
	    resultListPtr = Tcl_GetObjResult(interp);
 	    Tcl_ListObjAppendElement(interp, resultListPtr, afterPtr->commandPtr);
 	    Tcl_ListObjAppendElement(interp, resultListPtr, Tcl_NewStringObj(
 		(afterPtr->token == NULL) ? "idle" : "timer", -1));
	    Tcl_SetObjResult(interp, resultListPtr);
	    break;
	}
```

Here is the current Tcl code:
```
    case AFTER_INFO:
	if (objc == 2) {
            Tcl_Obj *resultObj = Tcl_NewObj();

	    for (afterPtr = assocPtr->firstAfterPtr; afterPtr != NULL;
		    afterPtr = afterPtr->nextPtr) {
		if (assocPtr->interp == interp) {
                    Tcl_ListObjAppendElement(NULL, resultObj, Tcl_ObjPrintf(
                            "after#%d", afterPtr->id));
		}
	    }
            Tcl_SetObjResult(interp, resultObj);
	    return TCL_OK;
	}
	if (objc != 3) {
	    Tcl_WrongNumArgs(interp, 2, objv, "?id?");
	    return TCL_ERROR;
	}
	afterPtr = GetAfterEvent(assocPtr, objv[2]);
	if (afterPtr == NULL) {
            const char *eventStr = TclGetString(objv[2]);

	    Tcl_SetObjResult(interp, Tcl_ObjPrintf(
                    "event \"%s\" doesn't exist", eventStr));
            Tcl_SetErrorCode(interp, "TCL","LOOKUP","EVENT", eventStr, NULL);
	    return TCL_ERROR;
	} else {
            Tcl_Obj *resultListPtr = Tcl_NewObj();

            Tcl_ListObjAppendElement(interp, resultListPtr,
                    afterPtr->commandPtr);
            Tcl_ListObjAppendElement(interp, resultListPtr, Tcl_NewStringObj(
		    (afterPtr->token == NULL) ? "idle" : "timer", -1));
            Tcl_SetObjResult(interp, resultListPtr);
        }
	break;
```
`objc == 2` is the difference between calling with an id and not calling with an id.

Here's the commit where the comment was added: https://github.com/python/cpython/commit/3c0f2c91adf086f809169f94ca9a2a75df3dc996 which changed the code:
```
-            (script, type) = self.tk.splitlist(
-                self.tk.call('after', 'info', id))
+            data = self.tk.call('after', 'info', id)
+            # In Tk 8.3, splitlist returns: (script, type)
+            # In Tk 8.4, splitlist may return (script, type) or (script,)
+            script = self.tk.splitlist(data)[0]
``` 

So, my thought is that somehow `self.tk.call('after', 'info', id))` was called with an id = None, which caused the info call to not return a 2-tuple.  I don't know how to look at the original issue report, so I don't know if it was able to be recreated or someone just patched the unpacking.

Reply:
As I remember, tkinter now requires at least 8.4, (while IDLE requires 8.5), so 8.3 is not relevant.